src/legacy.rs

```diff
@@ -0,0 +1,194 @@
+        let params = WebPushParams::new(rs, pad_length, salt);
+        let encrypted_block = AesGcmEceWebPush::encrypt_with_keys(
+            &*local_key_pair,
+            &*remote_pub_key,
+            &auth_secret,
+            &plaintext,
+            params,
+        )?;
+        Ok(AesGcmTestPayload {
+            dh: hex::encode(encrypted_block.dh),
+            salt: hex::encode(encrypted_block.salt),
+            rs: encrypted_block.rs,
+            ciphertext: hex::encode(encrypted_block.ciphertext),
+        })
+    }
+
+    fn try_decrypt(
+        private_key: &str,
+        public_key: &str,
+        auth_secret: &str,
+        payload: &AesGcmTestPayload,
+    ) -> Result<String> {
+        let private_key = hex::decode(private_key).unwrap();
+        let public_key = hex::decode(public_key).unwrap();
+        let ec_key = EcKeyComponents::new(private_key, public_key);
+        let plaintext = decrypt_aesgcm(
+            &ec_key,
+            &hex::decode(auth_secret).unwrap(),
+            &AesGcmEncryptedBlock::new(
+                &hex::decode(&payload.dh).unwrap(),
+                &hex::decode(&payload.salt).unwrap(),
+                payload.rs,
+                hex::decode(&payload.ciphertext).unwrap(),
+            )?,
+        )?;
+        Ok(String::from_utf8(plaintext).unwrap())
+    }
+
+    #[test]
+    fn test_e2e() {
+        let (local_key, remote_key) = crate::generate_keys().unwrap();
+        let plaintext = b"There was a green mouse, running in the grass";
+        let mut auth_secret = vec![0u8; 16];
+        let cryptographer = crate::crypto::holder::get_cryptographer();
+        cryptographer.random_bytes(&mut auth_secret).unwrap();
+        let remote_public = cryptographer
+            .import_public_key(&remote_key.pub_as_raw().unwrap())
+            .unwrap();
+        let params = WebPushParams::default();
+        let encrypted_block = AesGcmEceWebPush::encrypt_with_keys(
+            &*local_key,
+            &*remote_public,
+            &auth_secret,
+            plaintext,
+            params,
+        )
+        .unwrap();
+        let decrypted =
+            AesGcmEceWebPush::decrypt(&*remote_key, &auth_secret, &encrypted_block).unwrap();
+        assert_eq!(decrypted, plaintext.to_vec());
+    }
+
+    #[test]
+    fn test_conv_fn() -> Result<()> {
+        let (local_key, auth) = crate::generate_keypair_and_auth_secret()?;
+        let plaintext = b"There was a little ship that had never sailed";
+        let mut salt = vec![0u8; 16];
+        let cryptographer = crate::crypto::holder::get_cryptographer();
+        cryptographer.random_bytes(&mut salt)?;
+        let encoded = encrypt_aesgcm(&local_key.pub_as_raw()?, &auth, &salt, plaintext).unwrap();
+        let decoded = decrypt_aesgcm(&local_key.raw_components()?, &auth, &encoded)?;
+        assert_eq!(decoded, plaintext.to_vec());
+        Ok(())
+    }
+
+    #[test]
+    fn try_encrypt_ietf_rfc() {
```

Comment:
To cross-check, are these the same values from [the example in Section 5](https://tools.ietf.org/html/draft-ietf-webpush-encryption-04#section-5) of the linked RFC? It would be great if you could include a link in a comment here right the code for future readers to be able to find the source material.

Reply:
I took the example from draft 4, but they are indeed the same. I have just converted URL-safe Base64 to hexadecimal. It is doable for sure. I'll submit a few commits to handle those comments later today ;)